### PR TITLE
Chai Mocha tests

### LINKED ISSUE #58: Write tests
Since we have added more features to the application we need to update our tests as well

- Add E2E API tests
  For writing these tests we will be using [https://www.chaijs.com/plugins/chai-http/](chai-http)
  - [ ] Create `test/e2e` directory
  - [ ] For each of scenario we will have a separate file e.g. run scenario will have RunScenario.spec.ts
  - [ ] Write tests for run scenario
  Since there will be no consumers on the other end of the queue, we will have to call `onSuccess` method on the controllers manually using setTimeout function with the mock result and assert it with the response from the API
    - [ ] Run Code in Sync Mode and check if the response is correct
    - [ ] Run Code in Poll Mode
      - [ ] Check if the response is returning job id
      - [ ] Check if the job runs successfully after calling `onSuccess` method
    - [ ] Run Code in Callback Mode
      - [ ] Check if the response is returning job id
      - [ ] Setup another mock server or [echo-server](https://www.npmjs.com/package/http-echo-server) and assert the result
  - [ ] Write test for submit scenario
    - [ ] Run Code in Sync Mode and check if the response is correct
    - [ ] Run Code in Poll Mode
      - [ ] Check if the response is returning job id
      - [ ] Check if the job runs successfully after calling `onSuccess` method
    - [ ] Run Code in Callback Mode
      - [ ] Check if the response is returning job id
      - [ ] Setup another mock server or [echo-server](https://www.npmjs.com/package/http-echo-server) and assert the result 

- Add Unit tests
  - [ ] Create `test/unit` directory
  - [ ] Create test for Validators in each of the scenario
  - [ ] Write exhaustive set of test requests and assert the result

### PR DESCRIPTION
**Pull request by prabalsingh24**
fixes #58 
Added Chai and Mocha Tests

### CODE REVIEW

**Comment by prabalsingh24** [comment]
I changed the config file. Changed the DB to `judgeapi-tests`. So make sure you create another similar DB in postgres.
I think it would be better to have a `test-config.json`. and modify the `npm test` to use that config. Your thoughts?
@jatinkatyal13

**Comment by jatinkatyal13** [comment]
sounds good ! All the test config will be isolated from the main config. Just make sure to isolate them in the code as well.

**Comment by prabalsingh24** [comment]
I found some errors in the code. I've marked them as `TODO` in the tests. Should I go and open the issue or just solve them in this PR only.
I guess it'd be better if you have a look at them first ( I might be wrong :P)

Also out of curiosity, what do you mean by 'Referer' in api whitelist domains. [here](https://github.com/coding-blocks/judge-api/blob/72944d1764f56a6c06af7b1a46c90cfd16872e7d/src/validators/ApiKeyValidators.ts#L22)

And shouldn't it be `apiKey.whitelist_domains.indexOf('Referer') !==-1 `

File: test/unit/validators/RunValidator.spec.ts [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        const req = {
+            body: {
+                source: 'LKJSDFKLMC414CcnBcba12',
+                lang: 'cpp',
+                stdin: 'INPUT'
+            }
+        };
+
+        await runValidator.POST(req, res, nextSpy);
+        // TODO
```
Comment (jatinkatyal13) [inline test/unit/validators/RunValidator.spec.ts:104]: mode is allowed to be empty by design, as the default mode is sync.

File: test/unit/validators/RunValidator.spec.ts [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+                lang: 'cpp',
+                mode: 123,
+                stdin: 'INPUT'
+            }
+        };
+
+        await runValidator.POST(req, res, nextSpy);
+        // TODO
+        expect(nextSpy.calledOnce).to.be.false;
+        // error thrown == " mode" must be one of [sync, callback, poll]
```
Comment (jatinkatyal13) [inline test/unit/validators/RunValidator.spec.ts:123]: We are relying on Joi errors and as far as this is concerned, we may move to modes being numerical. For e.g. sync = 0, callback = 1

**Comment by jatinkatyal13** [comment]
Impressive pull request 🚀 

The entire purpose of having the tests is to have a code that works in the production environment. We can fix these errors in different PRs, but identifying them with the exhaustive set test is important.

Let's just Review the ToDos in the test and complete the PR

File: config.js [hunk 3]
```diff
@@ -10,7 +10,7 @@ exports = module.exports = {
 
   // Database config for a postgresql db (docker based or native)
   DB: {
-    DATABASE: process.env.DB_NAME || 'judgeapi',
+    DATABASE: process.env.DB_NAME || 'judgeapi-test',
```
Comment (jatinkatyal13) [inline config.js:13]: Let's use test-config

File: test/e2e/LangsScenario.spec.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+            whitelist_ips: ['*']
+        });
+    });
+    after(utils.truncateTables);
+
+
+    it('should throw 403 error API key is absent in the request', async () => {
+        const res = await chai.request(app).get(`/api/langs`);
+        expect(res.status).to.equal(403);
+        // TODO this should be res.body.err maybe? (consistent with other error)
```
Comment (jatinkatyal13) [inline test/e2e/LangsScenario.spec.ts:33]: Status code will define the error and let's stick with body.message

File: test/e2e/RunScenario.spec.ts [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+            lang_name: 'C++',
+            lang_version: '11'
+        });
+        await utils.setupMockQueue()
+    });
+    after(utils.truncateTables);
+
+
+    it('should throw 403 error if API Key is absent', async () => {
+        // shouldn't this be bad request 400?
```
Comment (jatinkatyal13) [inline test/e2e/RunScenario.spec.ts:50]: 400 represents Bad request, like unwanted keys in the body. Whereas Unauthorized is sent if we are missing authorization information.

Clients will straightaway know that they are missing Authorization header if they receive 403

File: test/e2e/LangsScenario.spec.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    });
+
+    it('should throw error if incorrect API key is present', async () => {
+        const res = await chai.request(app).get('/api/langs').set({
+            'Authorization': 'Bearer incorrectAPI-KEY',
+            Accept: 'application/json'
+        });
+        expect(res.status).to.equal(403);
+        expect(res.body.message).to.equal('Invalid API Key');
+        // maybe 400 error?
```
Comment (jatinkatyal13) [inline test/e2e/LangsScenario.spec.ts:44]: same as below

File: test/e2e/RunScenario.spec.ts [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+            mode: 'sync'
+        };
+
+        const res = await chai.request(app).post(`/api/runs`).set({
+            Authorization: 'Bearer 7718330d2794406c980bdbded6c9dc1d',
+            Accept: 'application/json'
+        }).send(params);
+
+        expect(res.status).to.equal(200);
+        // TODO ? does not throw error.
```
Comment (jatinkatyal13) [inline test/e2e/RunScenario.spec.ts:90]: It should fail with a 400 status code.

File: test/unit/validators/RunValidator.spec.ts [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+                lang: 'cpp',
+                mode: 'poll',
+                stdin: '',
+                timelimit: '123'
+            }
+        };
+
+        await runValidator.POST(req, res, nextSpy);
+
+        // TODO fix this. CODE is wrong
```
Comment (jatinkatyal13) [inline test/unit/validators/RunValidator.spec.ts:202]: change this test with `timelimie: 'abc'` and see if it works.

File: test/e2e/LangsScenario.spec.ts (hunk 4 above)
Comment (prabalsingh24) [reply test/e2e/LangsScenario.spec.ts:33]: But the base validator is sending `body.err`  [base-validator-error](https://github.com/coding-blocks/judge-api/blob/72944d1764f56a6c06af7b1a46c90cfd16872e7d/src/validators/baseValidator.ts#L16)

File: test/e2e/RunScenario.spec.ts (hunk 7 above)
Comment (prabalsingh24) [reply test/e2e/RunScenario.spec.ts:90]: I've created an issue for this.

File: test/unit/validators/RunValidator.spec.ts (hunk 8 above)
Comment (prabalsingh24) [reply test/unit/validators/RunValidator.spec.ts]: it worked :) `'123'` is a string too though :'(

**Comment by prabalsingh24** [comment]
@jatinkatyal13 Have a look at the PR when you're free :)

**Comment by prabalsingh24** [comment]
I need your help with test-config thing. 

I am not able to change the `process.env.NODE_ENV`

I changed npm test to
 `cross-env NODE_PATH=src node_modules/.bin/mocha --timeout 12000 --exit --require ts-node/register test/unit/validators/*.ts test/e2e/*.ts && cross-env NODE_ENV=test mocha  `

But the enviroment is not getting changed.


When I did this
```cross-env NODE_ENV=test mocha && NODE_PATH=src node_modules/.bin/mocha --timeout 12000 --exit --require ts-node/register test/unit/validators/*.ts test/e2e/*.ts```

the `process.env.NODE_ENV` was getting changed to test but throwing `No test files found` error

**Comment by prabalsingh24** [comment]
about 1400 lines of code :sleepy: 
I hope I haven't made many mistakes :sweat_smile:

File: config/config.js [hunk 9]
```diff
@@ -0,0 +1,9 @@
+const developmentConfig = require('./development-config');
+const testConfig = require('./test-config');
+
+if (process.env.NODE_ENV === 'test') {
```
Comment (jatinkatyal13) [inline config/config.js:4]: Directory structure should be as follows

- config/index.js
- config/development.js
- conifg/test.js
- config/production.js

All the common configs will be exported from index.js and overridable based on the NODE_ENV.

**Comment by jatinkatyal13** [comment]
We should run `DB.sync` once before starting the test. The test DB, in any case, will be discarded and hence we need not truncate it,

**Comment by prabalsingh24** [comment]
@jatinkatyal13 made the changes. Also I am doing `truncateTable` to make sure there no interference b/w the tests. Otherwise unexpected things might happen when something which was created in the previous tests is affecting the other tests

**Comment by jatinkatyal13** [comment]
Generally following is the process of setting up DB
1. Setup Tables and Sequences
2. Seed mock data (Read only data common for all tests)
3. Seed data per test
4. Run test
5. Truncate per test data

**Comment by prabalsingh24** [comment]
>     1. Setup Tables and Sequences

I think I am not doing this right now. I am assuming there is judgeapi-test db with tables

**Comment by prabalsingh24** [comment]
> Generally following is the process of setting up DB

@jatinkatyal13  I think I am also doing something similar. Waiting for a complete review :)

**Comment by jatinkatyal13** [comment]
Setup is perfect now!

Let's remove all the TODOs and let the test fail in the case of wrong result. If the lang is incorrect we should get 400 and if the domain is not whitelisted we should get 403

**Comment by prabalsingh24** [comment]
Yaay. This is probably my biggest PR . :)

**Pull request merged**